subprojects/core/src/test/groovy/org/gradle/api/internal/attributes/DefaultMutableAttributeContainerTest.groovy

```diff
@@ -30,6 +30,71 @@ import spock.lang.Specification
 class DefaultMutableAttributeContainerTest extends Specification {
     def attributesFactory = AttributeTestUtil.attributesFactory()
 
+    def "lazy attributes are evaluated in insertion order"() {
+        def container = new DefaultMutableAttributeContainer(attributesFactory)
+        def actual = []
+        def expected = []
+        (1..100).each { idx ->
+            def testAttribute = Attribute.of("test"+idx, String)
+            expected << idx
+            container.attributeProvider(testAttribute, Providers.<String>changing {
+                actual << idx
+                "value " + idx
+            })
+        }
+        expect:
+        container.asImmutable()
+        actual == expected
+    }
+
+    def "realizing the value of lazy attributes may cause other attributes to be realized"() {
+        def container = new DefaultMutableAttributeContainer(attributesFactory)
+        def firstAttribute = Attribute.of("first", String)
+        def secondAttribute = Attribute.of("second", String)
+        container.attributeProvider(firstAttribute, Providers.<String>changing {
+            // side effect is to evaluate the secondAttribute's value and prevent
+            // it from changing by removing it from the list of "lazy attributes"
+            container.getAttribute(secondAttribute)
+            "first"
+        })
+        container.attributeProvider(secondAttribute, Providers.of("second"))
+
+        expect:
+        container.asImmutable()
```

Comment:
Can we verify that this container has the expected contents?

Reply:
that's a good idea